ffmpeg/nodes.py

```diff
@@ -80,44 +97,80 @@ def __check_input_types(cls, stream_map, incoming_stream_types):
         for stream in list(stream_map.values()):
             if not _is_of_types(stream, incoming_stream_types):
                 raise TypeError('Expected incoming stream(s) to be of one of the following types: {}; got {}'
-                    .format(_get_types_str(incoming_stream_types), type(stream)))
+                                .format(_get_types_str(incoming_stream_types), type(stream)))
 
     @classmethod
     def __get_incoming_edge_map(cls, stream_map):
         incoming_edge_map = {}
         for downstream_label, upstream in list(stream_map.items()):
-            incoming_edge_map[downstream_label] = (upstream.node, upstream.label)
+            incoming_edge_map[downstream_label] = (upstream.node, upstream.label, upstream.selector)
         return incoming_edge_map
 
     def __init__(self, stream_spec, name, incoming_stream_types, outgoing_stream_type, min_inputs, max_inputs, args=[],
-            kwargs={}):
+                 kwargs={}):
         stream_map = get_stream_map(stream_spec)
         self.__check_input_len(stream_map, min_inputs, max_inputs)
         self.__check_input_types(stream_map, incoming_stream_types)
         incoming_edge_map = self.__get_incoming_edge_map(stream_map)
         super(Node, self).__init__(incoming_edge_map, name, args, kwargs)
         self.__outgoing_stream_type = outgoing_stream_type
 
-    def stream(self, label=None):
+        self.__incoming_stream_types = incoming_stream_types
+        self.__min_inputs = min_inputs
+        self.__max_inputs = max_inputs
+
+    def stream(self, label=None, select=None):
         """Create an outgoing stream originating from this node.
 
         More nodes may be attached onto the outgoing stream.
         """
-        return self.__outgoing_stream_type(self, label)
+        return self.__outgoing_stream_type(self, label, upstream_selector=select)
 
-    def __getitem__(self, label):
+    def __getitem__(self, item):
         """Create an outgoing stream originating from this node; syntactic sugar for ``self.stream(label)``.
+        It can also be used to apply a selector: e.g. node[0:"audio"] returns a stream with label 0 and
+        selector "audio", which is the same as ``node.stream(label=0, select="audio")``.
         """
-        return self.stream(label)
+        if isinstance(item, slice):
+            return self.stream(label=item.start, select=item.stop)
+        else:
+            return self.stream(label=item)
+
+    def _add_streams(self, stream_spec):
+        """Attach additional streams after the Node is initialized.
+        """
```

Comment:
This is odd - Nodes should be considered immutable. What's the intent of retroactively rewiring an already-created node?

Reply:
I think I see now. Mapping can be done either with a multi-stream output or using `.map`, and `.map` takes an existing `output` node and makes it a multi-output node.

Can we just make a new output node with the desired streams rather than mutating the existing one?

Making nodes mutable has a lot of widespread implications and opens up the complexity to spiral out of control - for example, downstream hashes would need to be recursively recalculated. (See [some of the benefits of immutable data-structures](http://www.yegor256.com/2014/06/09/objects-should-be-immutable.html)). I'd rather not make such a fundamental change unless absolutely necessary.

Reply:
> Can we just make a new output node with the desired streams rather than mutating the existing one?

I agree that this should be the best way to do it, I did it this way as I didn't know how to get back the stream specs from an existing node to regenerate the new OutputNode.

Do you have any ideas?